File: BackUpForABook/UIControls/TRClasses/TRMap.m
//
//  TRMap.m
//  ATreeRingMap
//
//  Created by Hui Wang on 5/20/14.
//  Copyright (c) 2014 AnnieLyticx. All rights reserved.
//

#import "TRMap.h"
#import "TRMLabels.h"

@interface TRMap (PrivateMethods)
// private methods
// preResize
-(void)preResize;

// off screen drawing
- (void)drawRect_Offscreen:(CGRect)rect;

// cached image context for off screen drawing
-(void)initContext:(CGSize)size;

// create image based CGContext
-(CGContextRef)createBitmapContext:(CGSize)size;

// paint background
-(void)paintBackground:(CGContextRef)ctx
                  size:(CGRect)rect;

// resize slice angle size
-(void)setSliceSize:(TRSlice *)slice
          sliceSize:(Slice) sliceSize;

// move slice to new origin
-(void)moveSlice:(TRSlice *)slice
          origin:(CGPoint)origin;

// move slice column slider bar
-(void)moveSliceColumnSliderBar:(TRSlice *)slice
                    barPosition:(float) barPosition;

// move slice row slider bar
-(void)moveSliceRowSliderBar:(TRSlice *)slice
                 barPosition:(float) barPosition;

// object hit test
-(void)hitTest:(CGPoint)hitPt;

@end

@implementation TRMap
{
    @private
    // private properties

    // cached image context
    CGContextRef _cacheContext;

    // touch events related
    // index of object being hitten
    int _idxObj;
    // hit object
    HitObj _hitObj;

    // begin touch point
    CGPoint _ptBegin;
    // end touch point
    CGPoint _ptEnd;
}

- (id)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    
    if (self == nil)
    {
        return nil;
    }
        
    // Initialization code
    self.backgroundColor = [UIColor whiteColor];

    _offscreen = FALSE;
    
    // cached image context
    if (_offscreen == TRUE)
    {
        _cacheContext = [self createBitmapContext:frame.size];
    }else
    {
        _cacheContext = nil;
    }

    _idxObj = -1;
    _hitObj.hitObject = TRObjs_None;

    // lables
    _lables = [[TRMLabels alloc] init];
    
    // notes
    _notes = [[TRMLabels alloc] init];

    // legend
    _legend = [[TRLegend alloc] init];
    
    // date
    _date = [[TRMLabels alloc] init];
    
    // center label
    _centerLabel = [[TRCenterLabel alloc] init];
    
    _fontSize = 1.0;
    _innerSpace = 70.0;
    _outerSpace = 100.0;
    
    _origin.x = _origin.y = 0;
    _min = .0;
    _max = .0;
     
    return self;
}

-(CGContextRef)createBitmapContext:(CGSize)size
{
    CGContextRef    context = NULL;
    CGColorSpaceRef colorSpace;
    void *          bitmapData;
    int             bitmapByteCount;
    int             bitmapBytesPerRow;
    
    // {r, g, b, alpha}
    bitmapBytesPerRow   = (size.width * 4);
    bitmapByteCount     = (bitmapBytesPerRow * size.height);
    
    colorSpace = CGColorSpaceCreateDeviceRGB();
    bitmapData = malloc( bitmapByteCount );
    if (bitmapData == NULL)
    {
        CGColorSpaceRelease(colorSpace);
        return nil;
    }
    
    colorSpace = CGColorSpaceCreateDeviceRGB();
    bitmapData = malloc( bitmapByteCount );
    if (bitmapData == NULL)
    {
        CGColorSpaceRelease(colorSpace);
        return nil;
    }
    
    context = CGBitmapContextCreate (bitmapData,
									 size.width,
									 size.height,
									 8,
									 bitmapBytesPerRow,
									 colorSpace,
									 kCGImageAlphaPremultipliedLast);   // kCGImageAlphaPremultipliedLast
    
    if (context== NULL)
    {
        free (bitmapData);
        CGColorSpaceRelease(colorSpace);
        return nil;
    }
    
    CGColorSpaceRelease( colorSpace );
    
    return context;
}

// Only override drawRect: if you perform custom drawing.

// An empty implementation adversely affects performance during animation.
- (void)drawRect:(CGRect)rect
{
    if (_offscreen == TRUE)
    {
        return [self drawRect_Offscreen:rect];
    }
 
    // regular drawing
    // Drawing code
    CGContextRef ctx = UIGraphicsGetCurrentContext();
    
    // (1) paint background
    [self paintBackground:ctx size:rect];
    
    if (allSlices.count == 0)
    {
        return ;
    }

    // (2) Draw all slices.
    int i;
    for (i = 0; i < allSlices.count; i++)
    {
        if ([[allSlices objectAtIndex:i] isKindOfClass:[TRSlice class]])
        {
            [[allSlices objectAtIndex:i] Paint:ctx];
        }
    }
 
    // (3) paint labels
    CGPoint pos, ltPosition = _origin;
    
    ltPosition.x = rect.origin.x + 10 * _fontSize;
    ltPosition.y -= _max;
    ltPosition.y = .5 * (rect.origin.y + ltPosition.y);
    
    pos = ltPosition;
    pos = [_lables paint:ctx orderByKeys:_lables.keys alignment:TRMTextAlignment_LT lineSpace:2.0 position:pos];
    
    // (4) paint legend
    CGPoint ltLegendPosition = pos;
    ltLegendPosition.y += 20;
    
    pos = [_legend paint:ctx orderByKeys:nil alignment:TRLegendAlignment_LT lineSpace:2.0 position:ltLegendPosition];
    
    // (5) paint date
    pos = ltPosition;
    pos.x = rect.origin.x + rect.size.width - 10 * _fontSize;
    
    [_date paint:ctx orderByKeys:_date.keys alignment:TRMTextAlignment_RT lineSpace:2.0 position:pos];
    
    // (6) paint note (on the right bottom)
    pos = ltPosition;
    pos.x = rect.origin.x + rect.size.width - 10 * _fontSize;
    pos.y = _origin.y + 80;     // !!! To DO.  Use the position of end of axis
    
    [_notes paint:ctx orderByKeys:_notes.keys alignment:TRMTextAlignment_RT lineSpace:2.0 position:pos];
    
    // (7) paint center label
    [_centerLabel paint:ctx radius:(_min - 10 * _fontSize) origin:_origin];
}

// off screen drawing
- (void)drawRect_Offscreen:(CGRect)rect
{
    // Drawing code
    CGContextRef ctx = UIGraphicsGetCurrentContext();
  
    // draw everythong onto the cached context
    // offscreen drawing
    // (1) paint background
    [self paintBackground:_cacheContext size:rect];
    
    // (2) Draw all slices.
    int i;
    for (i = 0; i < allSlices.count; i++)
    {
        if ([[allSlices objectAtIndex:i] isKindOfClass:[TRSlice class]])
        {
            [[allSlices objectAtIndex:i] Paint:_cacheContext];
        }
    }
    
    // done drawing on cached context
    // get image
    CGImageRef cacheImage = CGBitmapContextCreateImage(_cacheContext);
    // put the image onto current device context
    CGContextDrawImage(ctx, rect, cacheImage);
    
    // done
    CGImageRelease(cacheImage);
}

// touch events
-(void)touchesBegan:(NSSet *)touches withEvent:(UIEvent *)event
{
    UITouch *touch = [touches anyObject];
    
    // hit position
    _ptBegin = [touch locationInView:self];
    
    // object hit test
    [self hitTest:_ptBegin];
    
    // did we hit the axis ?
    TRSlice *hitSlice = nil;
    if (_idxObj >= 0 && (_hitObj.hitObject == TRObjs_Column_Axis || _hitObj.hitObject == TRObjs_Row_Axis))
    {
        // slice object
        hitSlice = [allSlices objectAtIndex:_idxObj];
        
        // Center position
        CGPoint center = hitSlice.origin;
        
        // {angle, radius} point position
        CGFloat r = hypotf(_ptBegin.x - center.x, _ptBegin.y - center.y);
        
        float endAngle = .0;
        endAngle = 180.0 * atan(-(_ptBegin.y - center.y) / (_ptBegin.x - center.x)) / 3.14;
        if ((_ptBegin.x - center.x) < 0)
        {
            endAngle += 180;
        }
        
        if (hitSlice == nil)
        {
            return ;
        }
        
        if (_hitObj.hitObject == TRObjs_Column_Axis)
        {
            // the column bar hit event should be handled inside the slice and ask for update display after
            [self moveSliceColumnSliderBar:hitSlice barPosition:endAngle];
            [self setNeedsDisplay];
        }else if (_hitObj.hitObject == TRObjs_Row_Axis)
        {
            // the row bar hit event should be handled inside the slice and ask for update display after
            [self moveSliceRowSliderBar:hitSlice barPosition:r];
            [self setNeedsDisplay];
        }
     }
}

-(void)touchesMoved:(NSSet *)touches withEvent:(UIEvent *)event
{
    UITouch *touch = [touches anyObject];
    _ptEnd = [touch locationInView:self];
    
    if (_idxObj < 0 || _hitObj.hitObject == TRObjs_None)
    {
        return ;
    }
    
    TRSlice *hitSlice = nil;
    
    // walking down the list of supported object types
    if (_hitObj.hitObject == TRObjs_Arrows_Begin ||
        _hitObj.hitObject == TRObjs_Arrows_End ||
        _hitObj.hitObject == TRObjs_ColumnBar ||
        _hitObj.hitObject == TRObjs_RowBar)
    {
        // slice object
        hitSlice = [allSlices objectAtIndex:_idxObj];
        
        if (hitSlice == nil)
        {
            return ;
        }
    
        // Center position
        CGPoint center = hitSlice.origin;
        
        // {angle, radius} point position
        CGFloat r = hypotf(_ptEnd.x - center.x, _ptEnd.y - center.y);
        
        float endAngle = .0;
        endAngle = 180.0 * atan(-(_ptEnd.y - center.y) / (_ptEnd.x - center.x)) / 3.14;
        if ((_ptEnd.x - center.x) < 0)
        {
            endAngle += 180;
        }
        
        Slice size = hitSlice.size;
        
        if (_hitObj.hitObject == TRObjs_Arrows_Begin)
        {
            // change the size of hitted slice first
            size.right = endAngle;
            [self setSliceSize:hitSlice sliceSize:size];
            
            // divider between two slices
            if (_idxObj >= 1)
            {
                hitSlice = [allSlices objectAtIndex:(_idxObj - 1)];
                size = hitSlice.size;
                size.left = endAngle;
                [self setSliceSize:hitSlice sliceSize:size];
            }
            
            [self autoFit];
            [self.layer setNeedsDisplay];
        }else if (_hitObj.hitObject == TRObjs_Arrows_End)
        {
            // change the size of hitted slice first
            size.left = endAngle;
            [self setSliceSize:hitSlice sliceSize:size];
            
            // divider between two slices
            if (_idxObj < ([allSlices count] - 1))
            {
                hitSlice = [allSlices objectAtIndex:(_idxObj + 1)];
                size = hitSlice.size;
                size.right = endAngle;
                [self setSliceSize:hitSlice sliceSize:size];
            }
            
            [self autoFit];
            [self.layer setNeedsDisplay];
        }else if (_hitObj.hitObject == TRObjs_ColumnBar)
        {
            // the column bar hit event should be handled inside the slice and ask for update display after
            [self moveSliceColumnSliderBar:hitSlice barPosition:endAngle];
            [self setNeedsDisplay];
        }else if (_hitObj.hitObject == TRObjs_RowBar)
        {
            // the row bar hit event should be handled inside the slice and ask for update display after
            [self moveSliceRowSliderBar:hitSlice barPosition:r];
            [self setNeedsDisplay];
        }
    }
    
    _ptBegin = _ptEnd;
}

-(void)touchesEnded:(NSSet *)touches withEvent:(UIEvent *)event
{
    [self setNeedsDisplay];
}

-(void)touchesCancelled:(NSSet *)touches withEvent:(UIEvent *)event
{
    _hitObj.hitObject = TRObjs_None;
    [self touchesEnded:touches withEvent:event];
}

// hit test walking down the object list. _idxObj will be -1 if hit no object
-(void)hitTest:(CGPoint)hitPt
{
    _idxObj = -1;
    
    // go over object (ring slices) list
    int i;
    for (i = 0; i < allSlices.count; i++)
    {
        if ([[allSlices objectAtIndex:i] isKindOfClass:[TRSlice class]])
        {
            _hitObj = [[allSlices objectAtIndex:i] HitTest:_ptBegin];
            
            if (_hitObj.hitObject != TRObjs_None)
            {
                _idxObj = i;
                _hitObj.sliceIndex = i;
                return ;
            }
        }
    }
}

// compose tree ring map by passing a set of tree slice angle positions.  usually called at initialization phase
-(void)composeTreeRingMap:(NSArray*)slicePositions
{
    if (slicePositions.count <= 1)
    {
        return ;
    }
    
    // clear allSlices first
    if (allSlices == nil)
    {
        allSlices = [[NSMutableArray alloc] init];
    }else
    {
        [allSlices removeAllObjects];
    }
    
    // creates tree slices
    int i;
    float b, e;
    TRSlice *sliceObj = nil;
    Slice sliceSize;
    
    // slice begin angle position
    b = [[slicePositions objectAtIndex:0] floatValue];
    
    for (i = 0; i < (slicePositions.count - 1); i++)
    {
        // slice end angle position
        e = [[slicePositions objectAtIndex:(i + 1)] floatValue];
        
        // create one slice object and add it to allSlices collections
        sliceObj = [[TRSlice alloc] init];
        
        sliceObj.parent = self.layer;
        // slice angle range (b, e)
        sliceSize = sliceObj.size;
        sliceSize.right = b;
        sliceSize.left = e;
        sliceObj.size = sliceSize;
        
        // only slices on left and right most has row lables
        if (i == 0)
        {
            // right most slice
            sliceObj.background.style = BackgroudStyle_Color_Fill;
            sliceObj.rowBarStyle = BarPosition_RB;
            sliceObj.rowLabelOnLeft = FALSE;
            sliceObj.rowLabelOnRight = TRUE;
            sliceObj.leftBorderStyle = SliceBorder_Line;
            sliceObj.rightBorderStyle = SliceBorder_Metal;
        }else if (i == (slicePositions.count - 2))
        {
            // left most slice
            sliceObj.background.style = BackgroudStyle_Color_Fill; //BackgroudStyle_Color_Gradient;
            sliceObj.rowBarStyle = BarPosition_RB; //BarPosition_LT;
            sliceObj.rowLabelOnLeft = TRUE;
            sliceObj.rowLabelOnRight = FALSE;
            sliceObj.leftBorderStyle = SliceBorder_Metal;
            sliceObj.rightBorderStyle = SliceBorder_Line;
        }else
        {
            sliceObj.background.style = BackgroudStyle_Color_Fill;
            sliceObj.rowBarStyle = BarPosition_None;
            sliceObj.rowLabelOnLeft = FALSE;
            sliceObj.rowLabelOnRight = FALSE;
            
            sliceObj.borderColor = [UIColor colorWithRed:1.0 green:1.0 blue:1.0 alpha:1.0];
            sliceObj.leftBorderStyle = SliceBorder_Line;    // SliceBorder_Solid;
            sliceObj.rightBorderStyle = SliceBorder_Line;   // SliceBorder_Solid;
        }
        
        // add it into collection allSlices
        [allSlices addObject:sliceObj];
        
        b = e;
    }
}

// get number of slices
-(int)numberOfSlices
{
    return allSlices.count;
}

// access to ring slice
-(TRSlice*)getSlice:(int)position
{
    if (allSlices == nil || allSlices.count == 0)
    {
        return nil;
    }
    
    return [allSlices objectAtIndex:position];
}

// set visual space for the tree ring map.  make sure the control is set with frame size first
-(void)setTRSize:(CGPoint)origin
             min:(float)min
             max:(float)max
{
    Slice sliceSize;
    TRSlice *oneSlice;
    
    int i;
    for (i = 0; i < allSlices.count; i++)
    {
        if( [[allSlices objectAtIndex:i] isKindOfClass:[TRSlice class]] )
        {
            oneSlice = [allSlices objectAtIndex:i];
            
            sliceSize = oneSlice.size;
            sliceSize.bottom = min;
            sliceSize.top = max;
            
            oneSlice.hostFrame = self.bounds;
            oneSlice.size = sliceSize;
            oneSlice.origin = origin;
        }
    }
    
    // make a note
    _origin = origin;
    _min = min;
    _max = max;
}

// auto resize the tree ring map
-(void)autoResize:(CGRect)size
{
    // set tree ring map size
    // _rmin (70) font point in inner circle area
    float rmin = _innerSpace * _fontSize;
    
    // maximum possible radius
    float rmax = .5 * size.size.width;
    
    // initially at the center location
    CGPoint origin = CGPointMake(size.size.width / 2, size.size.height / 2);
    
    // adjust ring size and position to best fit into given size.
    // adjustments in rmin, rmax and origin
    float d = 0.0;
    if ([self numberOfSlices] >= 1)
    {
        TRSlice* oneSlice = [self getSlice:0];
        if ((oneSlice.size.right < 0.0 || oneSlice.size.right > 180.0) &&
            ABS(sinf(DEGREES_TO_RADIANS(oneSlice.size.right))) > d)
        {
            d = ABS(sinf(DEGREES_TO_RADIANS(oneSlice.size.right)));
        }
        
        oneSlice = [self getSlice:([self numberOfSlices] - 1)];
        if ((oneSlice.size.left < .0 || oneSlice.size.left > 180.0) &&
            ABS(sinf(DEGREES_TO_RADIANS(oneSlice.size.left))) > d)
        {
            d = ABS(sinf(DEGREES_TO_RADIANS(oneSlice.size.left)));
        }
    }else
    {
        d = 1.0;
    }
    
    // adjustment in rmax
    if ((1.0 + d) * rmax > size.size.height)
    {
        rmax = size.size.height / (1.0 + d);
    }
    
    // set origin
    origin.y = size.origin.y + rmax;
    
    // shrink tree ring by _outerSpace (60 font points)
    float dR = _outerSpace * _fontSize;
    
    rmax -= dR;
    origin.y += .5 * dR;
    
    // center position, radius for bottom and top rings
    [self setTRSize:origin min:rmin max:rmax];
    
    // dirty view
    [self onDirtyView];
    
    // sync up slice size
    [self onSize];
}

-(void)autoFit
{
    [self autoResize:self.bounds];
}

// call to make sure tree ring slices are all in sync between view and data set
-(void)onDirtyView
{
    // (1) sync up with each slice components sucn as x and y axis
    for (int i = 0; i < [self numberOfSlices]; i++)
    {
        [[self getSlice:i] onDirtyView];
    }
}

-(void)onSize
{
    for (int i = 0; i < [self numberOfSlices]; i++)
    {
        [[self getSlice:i] onSize];
    }
}

// set slice size
-(void)setSliceSize:(TRSlice *)slice
          sliceSize:(Slice) sliceSize
{
    // hosting frame bound for coordinate conversion
    slice.hostFrame = self.bounds;
    // slice bound
    slice.size = sliceSize;
    [slice onDirtyView];
}

// move slice to new origin
-(void)moveSlice:(TRSlice *)slice
          origin:(CGPoint)origin
{
    slice.origin = origin;
}

// move slice column slider bar
-(void)moveSliceColumnSliderBar:(TRSlice *)slice
                    barPosition:(float) barPosition
{
    [slice setColumnAxisSliderBarPosition:barPosition];
}

// move slice row slider bar
-(void)moveSliceRowSliderBar:(TRSlice *)slice
                 barPosition:(float) barPosition
{
    [slice setRowAxisSliderBarPosition:barPosition];
}

// paint background
-(void)paintBackground:(CGContextRef)ctx
                  size:(CGRect)rect
{
    CGContextSetFillColorWithColor(ctx, self.backgroundColor.CGColor);
    CGContextFillRect(ctx, rect);
}

@end

File: BackUpForABook/UIControls/TRClasses_swift/TRMap.swift
//
//  TRMap.swift
//  TreeRingMap_Swift
//
//  Created by dingf on 17/3/11.
//  Copyright © 2017年 MH.dingf. All rights reserved.
//

// main components of trmap all in here: very important class


import UIKit

class TRMap: UIControl {
    // off Screen drawing
    var offScreen = false
    // index of obj being hit
    var idxObj: Int = -1
    var hitObj = HitObj()
    var fontSize: Float?
    var allSlices: [TRSlice]?
    var innerSpace: Float?
    var outerSpace: Float?
    var origin: CGPoint?
    var min: Float?
    var max: Float?
    // begin touch point
    var ptBegin: CGPoint?
    // end touch point
    var ptEnd: CGPoint?
    
    // label in the center
    var centerLabel: TRCenterLabel?
    // labels on left_top
    var labels: TRMLabels?
    // date label
    var date: TRMLabels?
    var lastScale: CGFloat = 1.0
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        backgroundColor = UIColor.clear//(red: 151.0/255.0, green: 190/255.0, blue: 76/255.0, alpha: 1.0)
        // 
        if offScreen == true{
        
        }else{
        
        }
        
        // labels 
        labels = TRMLabels.init()
        
        // center label
        centerLabel = TRCenterLabel.init()
        
        // date label
        date = TRMLabels.init()
        
        hitObj.hitObject = .none
        fontSize = 1.0
        innerSpace = 70.0
        outerSpace = 60.0
        origin = CGPoint(x: 0, y: 0)
        min = 0
        max = 0
        
        // pinch 
        let pinchRecognizer = UIPinchGestureRecognizer(target: self, action: #selector(pinchAction(_:)))
        addGestureRecognizer(pinchRecognizer)
        
    }
    
    func pinchAction(_ sender: UIPinchGestureRecognizer) {
        if sender.state == UIGestureRecognizerState.ended {
            lastScale = 1.0
            return
        }
        let scale = 1.0 - (lastScale - sender.scale)
        let currentTransform = self.transform
        let newTransform = currentTransform.scaledBy(x: scale, y: scale)
        self.transform = newTransform
        lastScale = sender.scale
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    func composeTreeRingMap(_ slicePositions: [NSNumber]){
        if slicePositions.count <= 1 {return}
        
        if allSlices == nil{
            allSlices = [TRSlice]()
        }else{
            allSlices?.removeAll()
        }
        
        // create tree slices
        var b = slicePositions[0].floatValue
        var e: Float
        
        for i in 0..<slicePositions.count - 1{
            e = slicePositions[i + 1].floatValue
            let sliceObj = TRSlice.init()
            sliceObj.parent = self.layer
            sliceObj.size.left = e
            sliceObj.size.right = b
            
            // right most slice
            if i == 0 {
                sliceObj.background!.style = .colorFill
                sliceObj.rowBarStyle = .rightOrBottom
                sliceObj.rowLabelOnLeft = false
                sliceObj.rowLabelOnRight = true
                sliceObj.leftBorderStyle = .line
                sliceObj.rightBorderStyle = .metal
            }else if i == (slicePositions.count - 2){
                sliceObj.background!.style = .colorFill
                sliceObj.rowBarStyle = .rightOrBottom
                sliceObj.rowLabelOnLeft = true
                sliceObj.rowLabelOnRight = false
                sliceObj.leftBorderStyle = .metal
                sliceObj.rightBorderStyle = .line
            }else{
                sliceObj.background!.style = .colorFill
                sliceObj.rowBarStyle = BarPosition.none
                sliceObj.rowLabelOnLeft = false
                sliceObj.rowLabelOnRight = false
                sliceObj.leftBorderStyle = .line
                sliceObj.rightBorderStyle = .line
                sliceObj.borderColor = UIColor(red: 1.0, green: 1.0, blue: 1.0, alpha: 1.0)
            }
            
            allSlices?.append(sliceObj)

            b = e
        
        }
      
    }
    
    func autoResize(_ size: CGRect){
        let rmin = innerSpace! * fontSize!
        var rmax: Float = 0.5 * Float(size.size.width)
        var origin  = CGPoint(x: size.size.width / 2.0, y: size.size.height / 2.0)
        var d: Float = 0.0
        if allSlices!.count >= 1{
            let oneSlice = allSlices![0]
            let value = abs(sinf(DEGREETORADIANS(degree: oneSlice.size.right!))) > d
            if (oneSlice.size.right! < Float(0.0) || oneSlice.size.right! > Float(180.0)) && (value == true){
                d = abs(sinf(DEGREETORADIANS(degree: oneSlice.size.right!)))
            }
            let lastSlice = allSlices!.last
            let value1 = abs(sinf(DEGREETORADIANS(degree: oneSlice.size.left!))) > d
            if ((lastSlice?.size.left!)! < Float(0.0) || (lastSlice?.size.left!)! > Float(180.0)) && (value1 == true){
                d = abs(sinf(DEGREETORADIANS(degree: (lastSlice?.size.left!)!)))
            }
        }else{
            d = 1.0
        }
        
        if (1.0 + d) * Float(rmax) > Float(size.size.height){
            rmax = Float(size.size.height) / (1.0 + d)
        }
        
        origin.y = size.origin.y + CGFloat(rmax)
        let dR: Float = outerSpace! * fontSize!
        rmax -= dR
        origin.y += 0.5 * CGFloat(dR)
        setTRsize(origin , min: rmin, max: rmax)
        onDirtyView()
        onSize()
      
    }
    
    func setTRsize(_ origin: CGPoint, min: Float, max: Float){
        for i in 0..<allSlices!.count{
            allSlices![i].hostFrame = bounds
            allSlices![i].origin = origin
            allSlices![i].size.bottom = min
            allSlices![i].size.top = max
          
        }
        self.origin = origin
        self.min = min
        self.max = max
    }
    
    func onDirtyView(){
        for i in 0..<allSlices!.count{
            allSlices![i].onDirtyView()
        }
        
    }
    
    func onSize(){
        for i in 0..<allSlices!.count{
            allSlices![i].onSize()
        }
    }
   
    // MARK: draw begin
    override func draw(_ rect: CGRect) {
        let ctx = UIGraphicsGetCurrentContext()
        // 1. paint background
        ctx?.setFillColor(backgroundColor!.cgColor)
        ctx?.fill(rect)
        
        // 2. draw all slices
        for i in 0..<allSlices!.count{
            self.allSlices?[i].paint(ctx!)
        }
        
        // 3. paint center label
        centerLabel!.paint(ctx!, radius: (min! - 10 * fontSize!), origin: origin!,borderWidth: 10 * fontSize!)
        
        // 4. paint labels on left_top
        let ltPosition = CGPoint(x: rect.origin.x + CGFloat(10 * fontSize!), y: 0.5 * (rect.origin.y + origin!.y - CGFloat(max!)))
       
        var pos = labels?.paint(ctx!, orderByKeys: labels!.keys!, alignment: TRMTextAlignment.left_top, lineSpace: 2.0, position: ltPosition)
        
        // 5. paint date label
        pos = ltPosition
        pos?.x = rect.origin.x + rect.size.width - 10 * CGFloat(fontSize!)
        print("-----note: pos\(date?.keys)-----")
        let _ = date?.paint(ctx!, orderByKeys: date!.keys!, alignment: TRMTextAlignment.right_top, lineSpace: 2.0, position: pos!)
    }
    
    // MARK: touch events
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        let touch = touches.first
        
        // hit position 
        ptBegin = touch?.location(in: self)
        
        // object hit test
        hitTest(ptBegin!)
        
        // did we hit the axis?
        var hitSlice: TRSlice
        if idxObj >= 0 && (hitObj.hitObject == TRObjs.column_axis || hitObj.hitObject == TRObjs.row_axis){
            // slice object 
            hitSlice = allSlices![idxObj]
            
            // center position
            let center = hitSlice.origin
            
            // {angle, radius} point position
            let r = hypotf(Float(ptBegin!.x - center!.x), Float(ptBegin!.y - center!.y))
            var endAngle = 180.0 * atan(-(ptBegin!.y - center!.y) / (ptBegin!.x - center!.x)) / 3.14
            if ptBegin!.x - center!.x < 0 {
                endAngle += 180
            }
            
            if hitObj.hitObject == TRObjs.column_axis{
                // the column bar hit event should be handled inside the slice and ask for update display after
                moveSliceColumnSliderBar(hitSlice, barPositon: Float(endAngle))
                setNeedsDisplay()
            }else if hitObj.hitObject == TRObjs.row_axis{
                // the row bar hit event should be handled inside the slice and ask for update display after
                moveSliceRowSliderBar(hitSlice, barPosition: r)
                setNeedsDisplay()
            }
        }
    }
    
    override func touchesMoved(_ touches: Set<UITouch>, with event: UIEvent?) {
        let touch = touches.first
        ptEnd = touch?.location(in: self)
        
        if idxObj < 0 || hitObj.hitObject == TRObjs.none{return}
        
        var hitSlice: TRSlice
        if hitObj.hitObject == TRObjs.arrows_begin || hitObj.hitObject == TRObjs.arrows_end || hitObj.hitObject == TRObjs.columnBar || hitObj.hitObject == TRObjs.rowBar{
            hitSlice = allSlices![idxObj]
            // center position
            let center = hitSlice.origin
            
            // {angle, radius} point position
            let r = hypotf(Float(ptEnd!.x - center!.x), Float(ptEnd!.y - center!.y))
            
            var endAngle: Float = 180.0 * Float(atan(-(ptEnd!.y - center!.y) / (ptEnd!.x - center!.x))) / Float(3.14)
            if (ptEnd!.x - center!.x) < 0 {
                endAngle += 180
            }
            var size = hitSlice.size
            
            if hitObj.hitObject == TRObjs.arrows_begin{
            
                // change the size of hitted slice first 
                size.right = endAngle
                setSliceSize(hitSlice, sliceSize: size)
                
                // divider between two slices
                if idxObj >= 1{
                    hitSlice = allSlices![idxObj - 1]
                    size = hitSlice.size
                    size.left = endAngle
                    setSliceSize(hitSlice, sliceSize: size)
                }
                autoResize(bounds)
                layer.setNeedsDisplay()
            }else if (hitObj.hitObject == TRObjs.arrows_end){
                // change the size of hitted slice first
                size.left = endAngle
                setSliceSize(hitSlice, sliceSize: size)
                
                // divider between two slices
                if idxObj < allSlices!.count - 1{
                    hitSlice = allSlices![idxObj + 1]
                    size = hitSlice.size
                    size.right = endAngle
                    setSliceSize(hitSlice, sliceSize: size)
                }
                autoResize(bounds)
                layer.setNeedsDisplay()
            }else if hitObj.hitObject == TRObjs.columnBar{
                moveSliceColumnSliderBar(hitSlice, barPositon: endAngle)
                setNeedsDisplay()
            }else if hitObj.hitObject == TRObjs.rowBar{
                // the row bar hit event should be handled inside the slice and ask for update display after
                moveSliceRowSliderBar(hitSlice, barPosition: r)
                setNeedsDisplay()
            }
            
        }
        ptBegin = ptEnd
        
    }
    
    override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
        setNeedsDisplay()
    }
    
    override func touchesCancelled(_ touches: Set<UITouch>, with event: UIEvent?) {
        hitObj.hitObject = TRObjs.none
        touchesEnded(touches, with: event)
    }
    
    // hit test will walk down the object list . idxObj will be -1 if hit no object
    fileprivate func hitTest(_ hitPt: CGPoint){
        idxObj = -1
        
        // go over object list(ring slices)
        for i in 0..<allSlices!.count{
            hitObj = allSlices![i].hitTest(ptBegin!)
            
            if hitObj.hitObject != TRObjs.none{
                idxObj = i
                hitObj.sliceIndex = i
                return
            }
        }
        
    }
    
    // move slice column slider bar
    func moveSliceColumnSliderBar(_ slice: TRSlice, barPositon: Float){
        slice.setColumnAxisSliderBarPosition(barPositon)
    }
    
    // move slice row slider bar
    func moveSliceRowSliderBar(_ slice: TRSlice, barPosition: Float){
        slice.setRowAxisSliderBarPosition(barPosition)
    }
    
    // set slice size
    func setSliceSize(_ slice: TRSlice, sliceSize: Slice){
        // hosting frame bound for coordinate conversion
        slice.hostFrame = self.bounds
        slice.size = sliceSize
        slice.onDirtyView()
    }
     
    
}
